### DISCUSSION TITLE
Long query execution time and missing results with back relation filter

### DISCUSSION THREAD

**Discussion by international-omelette** [pocketbase/pocketbase#5221] [Q&A]
I am running PocketBase v0.22.12 and I'm having an issue when applying a back relation filter to a table query. 

First, I will give some details about my case. These are my tables:
- `places` -> storing information about a community / institution
- `editions` -> representing an edition of a book in a certain `year`
- `submissions` -> representing form submission data of a `place` for an `edition`

The `submissions` table has these single relation columns:
- `place` -> links to the `places` table
- `edition` -> links to the `editions` table

For example, I want to filter all `places` that "have" a `submission` of a specific `edition.year`. So I tried to apply this filter on the `places` table:

```submissions_via_place.edition.year='2024'```

Unfortunately, this runs into a timeout (the request returns 400 after 30s). When I am reducing the depth of the filter, the query execution time decreases and the request can be resolved:

```submissions_via_place.edition='8nmd5g8zk05cnfx'```

However, the query takes >10s and when I run this in the admin panel, there's even a second request that fails:
<img width="1512" alt="Bildschirmfoto 2024-07-16 um 06 15 56" src="https://github.com/user-attachments/assets/a5939c35-3a33-4976-b2a1-fc07ab581289">

On further investigation, I found another example of back relation filtering not working properly or at least not how I expected it to work. There's another table for `publications`, storing information about the publishing details of a place in an edition. So the `publication` table also has a single type relation fields to a `place` and an `edition`.

To find out, which places are published in a specific edition, I applied this filter to the places table:

```publications_via_place.edition='8nmd5g8zk05cnfx'```

This returns 2 items, again with one fast and one slow response:
<img width="1512" alt="Bildschirmfoto 2024-07-16 um 06 41 11" src="https://github.com/user-attachments/assets/d375e2ae-14e3-49bf-a527-9c109e2b9c72">

But, checking the `publications` table, there are many more than 2 items with that edition (`8nmd5g8zk05cnfx` is the id of the _2024_ edition):
<img width="1512" alt="Bildschirmfoto 2024-07-16 um 06 42 13" src="https://github.com/user-attachments/assets/07d13281-b1a1-4174-85c2-0cca5b362d71">

Assuming I have correctly applied the back relation filter into my queries, this comes down to two issues:
- Long query duration
- Missing results

Regarding the query duration, I find it interesting that there's a significant gap between `submissions_via_place.edition` (took >10s to resolve) and `publication_via_place.edition` (took <200ms to resolve). Both tables have a similar amount of rows. But the submissions table is considerably bigger, with more columns, including quite a big json column that holds a snapshot of the complete `place` dataset at the time the `submission` was created. 

I hope I was able to provide all information necessary to start a valuable discussion around this topic.

**Comment by ganigeorgiev** [comment]
Unfortunately there is not enough details about your collections structure and your data set to reproduce the problems (ex. do you have any custom indexes?).

Regular relations filtering and (also back-relations) performs SQL JOINs and the query execution will highly depend on your data set.

As mentioned in the [Back-relations caveats](https://pocketbase.io/docs/working-with-relations/#back-relation-caveats) (at the bottom of the page) back-relations for most of the cases resolve as a **multiple `relation` field**, meaning that the filter `submissions_via_place.edition.year='2024'` will perform a **match-all** constraint, aka. this query could be slow because it will return `places` that have **all** of their `submissions`, and for each submission, **all** of their `editions` matching your constraint.

If you want to retrieve all places that has at least one submission and edition matching your constraint then you need to prefix the operator with `?`, aka:

`submissions_via_place.edition.year ?= '2024'`

_(this could explain also why you have missing results)_

To improve the performance of your query you can also try adding indexes for your relation fields (_avoid using composite indexes if you are not sure that they help as often they can have negative impact on the performance, ex. https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045_).

To assist you with the query debugging you can also try to start the executable with `--dev` and inspect the generated query.

If you need further help, please export your collections structure as json (Admin UI > Settings > Export collections) and add info on how many records each table has to that I can try to reproduce it locally.

**Reply by ganigeorgiev** [reply]
I've tried to simulate your setup based on the provided tables summary locally with 5k records in each collection and it returns the expected results for me (with the `?=` operator it responds under < 500ms).

But again, it will highly depend on your actual data set and what other columns your collections have. Most often the query could be slow because of:

- lack of indexes on the relation fields
- or the opposite - unnecessary indexes (often multi-column/composite) that can lead to inefficient query planner route (ex. https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045)
- large text/json field data per single record (again https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045)
- joins between large data sets due to the large cartesian product result that will have to be filtered
- multiple `relation` field "match-all" type of constraints (aka. without the `?` prefix) as this internally works by either generating a `COUNT` subquery or a `NOT EXISTS` subquery with the opposite constraints, which could lead to full table `SCAN`

**Reply by international-omelette** [reply]
Thanks for taking the time to review my request. To complete my setup description: there have not been any indexes set up.

Applying `?=` as the filter operator heavily speeds up the query execution time (<500ms) which is more than sufficient for my use case of finding places that _have at least one submission with a specific edition_.

However, if I try to find places that _do not have any submission with a specific edition_ by using the operator `?!=` the query execution time goes up to >20s. It eventually resolves and the results seem ok, though.

I have already set up indexes on the `place` and `edition` column of the `submissions` table. Is there anything else I could do to speed up the latter query?

If it is of interest I'll share more insights into the collection structure and data set of the tables involved.

**Reply by ganigeorgiev** [reply]
Again, without details about your data set and what other fields your collections have, there is nothing much I can help with as I have no way to reproduce your problem.

You can try to run the executable with `--dev` and transform the dev query to a plain SQLite one and run the sqlite3 cli against the `pb_data/data.db` file with an expression like `EXPLAIN QUERY PLAN yourQuery` to investigate what route the query planner has taken and whether an index could help.

In any case I suspect your problem is similar to the one in https://github.com/pocketbase/pocketbase/discussions/3859#discussioncomment-7744045 and that your collection has large text/json data fields, so one possible solution could be to split up the large fields that are not frequently queried into subcollections.
